### PR TITLE
pjsip: add darwin support

### PR DESCRIPTION
**Pull request by bobvanderlinden**
###### Description of changes

Currently the pjsip package is marked as broken for darwin because it is not able to build correctly.

This PR attempts to fix this for both aarch64-darwin and x86_64-darwin.

There are 2 workarounds in this PR:

* `fixDarwinDylibNames` wasn't able to rewrite the relative paths, because they contained `../` and `../../`. A similar method to `fixDarwinDylibNames` was used to rewrite them, but with support for `../` and `../`.
* On x86_64-darwin the library was automatically configured to use a deprecated `darwin-ssl` library, instead of openssl. This wasn't the case on aarch64, probably because `darwin-ssl` wasn't supported for aarch64 anymore. The configure flag `--disable-darwin-ssl` was needed to force using openssl on both architectures.

Lastly, because the linking issues were a bit finicky, I also added a package test to check whether the python library can be loaded correctly.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bobvanderlinden** [comment]
Is there a good way to let ofborg run the test (`pjsip.tests.python-pjsua2`) on all supported platforms?

File: pkgs/applications/networking/pjsip/default.nix [hunk 1]
```diff
@@ -41,21 +45,28 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ openssl libsamplerate ]
     ++ lib.optional stdenv.isLinux alsa-lib
-    ++ lib.optional stdenv.isDarwin AppKit;
+    ++ lib.optionals stdenv.isDarwin [ AppKit CoreFoundation Security ];
 
-  preConfigure = ''
-    export LD=$CC
-  '';
+
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:50]:
```suggestion
```

File: pkgs/applications/networking/pjsip/default.nix [hunk 2]
```diff
@@ -41,21 +45,28 @@ stdenv.mkDerivation rec {
...
+  NIX_CFLAGS_LINK = lib.optionalString stdenv.isDarwin
+                      "-headerpad_max_install_names";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:52]:
```suggestion
  NIX_CFLAGS_LINK = lib.optionalString stdenv.isDarwin "-headerpad_max_install_names";
```

File: pkgs/applications/networking/pjsip/default.nix [hunk 3]
```diff
@@ -41,21 +45,28 @@ stdenv.mkDerivation rec {
...
     make -C pjsip-apps/src/swig/python
   '';
 
+  preConfigure = ''
+    export LD=$CC
+  '';
+
+  configureFlags = [ "--enable-shared" ]
+    # darwin-ssl is a deprecated SSL library that requires
+    # a more recent version of Apple SDK 10 which is not in nixpkgs.
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:64]: That is wrong, see https://github.com/NixOS/nixpkgs/blob/54b3a83b4483d34daec601256769889bb11e0e81/doc/stdenv/platform-notes.chapter.md?plain=1#L66 how to use the newer sdk

File: pkgs/applications/networking/pjsip/default.nix [hunk 4]
```diff
@@ -65,13 +76,48 @@ stdenv.mkDerivation rec {
...
+        install_name_tool -id $lib "''${change_args[@]}" $lib
+      fi
+    done
   '';
 
   # We need the libgcc_s.so.1 loadable (for pthread_cancel to work)
   dontPatchELF = true;
 
+  passthru.tests.python-pjsua2 = runCommand "python-pjsua2" { } ''
+    ${python3.withPackages (pkgs: [ pkgs.pjsua2 ])}/bin/python -c "import pjsua2" > $out
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:117]:
```suggestion
    ${python3.withPackages (pkgs: [ pkgs.pjsua2 ]).interpreter} -c "import pjsua2" > $out
```

File: pkgs/applications/networking/pjsip/default.nix [hunk 5]
```diff
@@ -65,13 +76,48 @@ stdenv.mkDerivation rec {
...
+      lib_name="$(basename $lib)"
+      change_args+=(-change ../lib/$lib_name $lib)
+      change_args+=(-change ../../lib/$lib_name $lib)
+    done
+
+    # Rewrite id and library refences for all non-symlinked libraries.
+    for lib in "''${libraries[@]}"
+    do
+      if [ -f "$lib" ]
+      then
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:107]:
```suggestion
    for lib in "''${libraries[@]}"; do
      if [ -f "$lib" ]; then
```

File: pkgs/applications/networking/pjsip/default.nix [hunk 6]
```diff
@@ -65,13 +76,48 @@ stdenv.mkDerivation rec {
...
+  '' + lib.optionalString stdenv.isDarwin ''
+    # On MacOS relative paths are used to refer to libraries. All libraries use
+    # a relative path like ../lib/*.dylib or ../../lib/*.dylib. We need to
+    # rewrite these to use absolute ones.
+
+    # First, find all libraries (and their symlinks) in our outputs to define
+    # the install_name_tool -change arguments we should pass.
+    readarray -t libraries < <(
+      for outputName in $(getAllOutputNames)
+      do
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/pjsip/default.nix:88]:
```suggestion
      for outputName in $(getAllOutputNames); do
```

File: pkgs/applications/networking/pjsip/default.nix (hunk 1 above)
Comment (bobvanderlinden) [reply pkgs/applications/networking/pjsip/default.nix:52]: Removing this is unfortunately not possible. I've only moved it. When I tried to remove this I got:

```
       > /nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40/bin/ld: unrecognized option '-Wl,-soname,libpj.so.2'
       > /nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40/bin/ld: use the --help option for usage information
```

File: pkgs/applications/networking/pjsip/default.nix (hunk 3 above)
Comment (bobvanderlinden) [reply pkgs/applications/networking/pjsip/default.nix]: Thanks. That was the missing link. Works great! :smile:

**Comment by bobvanderlinden** [comment]
@SuperSandro2000 Thanks a lot for the review! Especially darwin's callPackage worked great :+1: I just couldn't remove the preConfigure.

**Pull request merged**